File: sandbox-platform/internal/router/router.go
package router

import (
	"context"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"github.com/sandbox/platform/internal/queue"
	"github.com/sandbox/platform/pkg/types"
)

type Router struct {
	mu    sync.RWMutex
	rules map[string]types.Tier
	qc    *queue.Client
}

func New(rdbClient *queue.Client) *Router {
	r := &Router{
		rules: make(map[string]types.Tier),
		qc:    rdbClient,
	}
	for tool, tier := range defaultRules() {
		r.rules[tool] = tier
	}
	return r
}

func (r *Router) Resolve(tool string) types.Tier {
	r.mu.RLock()
	defer r.mu.RUnlock()
	if tier, ok := r.rules[tool]; ok {
		return tier
	}
	return types.TierWASM // Default to WASM
}

func (r *Router) Register(tool string, tier types.Tier) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.rules[tool] = tier
}

// Execute routes and executes a job by pushing it to the required queue, and waits for a result.
func (r *Router) Execute(ctx context.Context, job types.Job) (types.RuntimeResult, error) {
	tier := r.Resolve(job.Tool)
	job.Tier = tier

	slog.Info("routing execution", "tool", job.Tool, "tier", tier, "job_id", job.ID)
	
	// Push to redis
	if err := r.qc.PushJob(ctx, job); err != nil {
		return types.RuntimeResult{Stderr: "failed to push job", ExitCode: 1}, fmt.Errorf("queue push: %w", err)
	}

	// Wait for response via redis
	res, err := r.qc.WaitForJobResult(ctx, job.ID, 30*time.Second)
	if err != nil {
		return types.RuntimeResult{Stderr: err.Error(), ExitCode: 1}, fmt.Errorf("wait result: %w", err)
	}

	return *res, nil
}

File: src/load-balancer/router/router.py
"""Job router: resolves tool → tier and pushes jobs to Redis.

Mirrors internal/router/router.go.
"""

from __future__ import annotations

import threading

import structlog

from load_balancer.queue_client import Client as QueueClient
from load_balancer.router.rules import default_rules
from load_balancer.types import Job, RuntimeResult, Tier

log = structlog.get_logger()


class Router:
    """Routes tools to execution tiers and dispatches jobs via the queue."""

    def __init__(self, queue_client: QueueClient) -> None:
        self._lock = threading.RLock()
        self._rules: dict[str, Tier] = dict(default_rules())
        self._qc = queue_client

    def resolve(self, tool: str) -> Tier:
        """Return the tier for tool, defaulting to WASM."""
        with self._lock:
            return self._rules.get(tool, Tier.WASM)

    def register(self, tool: str, tier: Tier) -> None:
        """Add or override a tool → tier mapping."""
        with self._lock:
            self._rules[tool] = tier

    def execute(self, job: Job, timeout: float = 30.0) -> RuntimeResult:
        """Route the job to its tier queue and wait for the result."""
        tier = self.resolve(job.tool)
        job.tier = tier

        log.info("routing execution", tool=job.tool, tier=tier.value, job_id=job.id)

        self._qc.push_job(job)

        try:
            return self._qc.wait_for_job_result(job.id, timeout=timeout)
        except TimeoutError as exc:
            return RuntimeResult(stderr=str(exc), exit_code=1)
